### ISSUE TITLE
Web STOMP: ERROR frame does not reach the client when subscribing to a non-existent target

### ISSUE THREAD

**Issue by csicar** [rabbitmq/rabbitmq-server#6737]
# Steps to Reproduce

1. Run RabbitMQ with `rabbitmq-plugins enable --offline rabbitmq_stomp rabbitmq_web_stomp`
2. Connect WebSocket to `ws://localhost:15674/ws`
3. `SUBSCRIBE` to non-existing exchange `/exchange/doesnotexist`

A minimal working example can be found here: https://gist.github.com/csicar/3452f6ab360e4401cddadeed76b7151d

# Expected Behaviour

As per [STOMP specification](https://stomp.github.io/stomp-specification-1.2.html#SUBSCRIBE)

> If the server cannot successfully create the subscription, the server MUST send the client an ERROR frame and then close the connection.

# Actual Behaviour

RabbitMQ closes the websocket connection before sending an `ERROR` frame

----

## Additional Information

Interestingly, the rabbitmq log indicates that an `ERROR` frame is sent, when in fact none is sent: https://gist.github.com/csicar/3452f6ab360e4401cddadeed76b7151d#file-rabbitmq-log-txt
>  STOMP error frame sent:

**Comment by michaelklishin** [comment]
If it sent but the connection is immediately closed, whether it reaches the client depends on relatively obscure socket settings such as `SO_LINGER`:

``` ini
stomp.tcp_listen_options.linger.on      = true
stomp.tcp_listen_options.linger.timeout = 0
```

that Web STOMP does not support via `rabbitmq.conf` and possibly cannot easily do so,
depending on what the WebSocket connection server it relies on exposes.

Use a traffic capture to confirm whether an `ERROR` frame is sent.

**Comment by michaelklishin** [comment]
It would also be curious to compare this against other targets.

**Comment by michaelklishin** [comment]
What does `rabbitmq-diagnostics environment` return? Specifically what is the [`linger` setting](https://stackoverflow.com/questions/3757289/when-is-tcp-option-so-linger-0-required) of the Web STOMP listener?

**Comment by csicar** [comment]
> Use a traffic capture to confirm whether an ERROR frame is sent.

I should have probably included that in the issue, but no: wireshark shows that directly after the client sends a `SUBSCRIBE`, RabbitMQ closes the websocket connection. See frame 16 (SUBSCRIBE frame) followed by frame 17 (CLOSE): https://gist.github.com/csicar/3452f6ab360e4401cddadeed76b7151d/#file-network-dump-log

----

> It would also be curious to compare this against other targets.

I tested the STOMP (i.e. not WebSTOMP) target with the same RabbitMQ-Configuration and it works as expected:
1. Client sends: SUBSCRIBE `/exchange/doesnotexist`
2. Server responds:  `Error: not_found`, followed by a TCP `FIN, ACK`

(see https://gist.github.com/csicar/3452f6ab360e4401cddadeed76b7151d#file-stomp-js)


----

> What does rabbitmq-diagnostics environment return?

See https://gist.github.com/csicar/3452f6ab360e4401cddadeed76b7151d#file-rabbitmq-diagnostics-txt . 
The WebSTOMP section does not contain any `linger` related information:
```
 {rabbitmq_web_stomp,
     [{cowboy_opts,[]},
      {num_ssl_acceptors,10},
      {num_tcp_acceptors,10},
      {proxy_protocol,false},
      {ssl_config,[]},
      {tcp_config,[{port,15674}]},
      {use_http_auth,false},
      {ws_frame,text}]},
````

**Comment by michaelklishin** [comment]
OK, thank for these details, they really help.

@lhoguin the issue may be in the plugin code (I'm happy to investigate) but how should we go about configuring `SO_LINGER` for WebSocket connections with Cowboy? Is there anything
less than obvious?

**Comment by lukebakken** [comment]
Thanks for all the info. I have this on my very long TODO list to follow up.

**Comment by csicar** [comment]
Well, I might be able to put it near the top of my TODO list to try to fix this.

Any pointers on where to get started?

**Comment by csicar** [comment]
I found a workaround (https://github.com/csicar/rabbitmq-server/tree/fix/web_stomp_missing_subscription), but I would like to understand further why RabbitMQ closes the STOMP session before sending the error message in the first place.

It that for reasons of convenience or security?

**Comment by lukebakken** [comment]
> It that for reasons of convenience or security?

Most likely it's a bug. I have yet to find time to investigate.

**Comment by michaelklishin** [comment]
Resolved in #6789.
